File: SearchingAndSorting/LastAndFirst/LastAndFirst.js
function method1(arr, start, end, n) {
  let first, last;
  let ind = binarySearch(arr, start, end, n);
  if (ind > 0 && arr[ind - 1] == n)
    first = binarySearch(arr, start, ind - 1, n, true, false);
  else first = ind;
  if (ind < arr.length - 1 && arr[ind + 1] == n)
    last = binarySearch(arr, ind + 1, end, n, false, true);
  else last = ind;
  return {
    first,
    last,
  };
}

function binarySearch(arr, start, end, n, isStart, isEnd) {
  let mid;
  while (start <= end) {
    mid = parseInt(start + (end - start) / 2);
    if (arr[mid] == n) {
      if (isStart) {
        if (mid > 0 && arr[mid - 1] == n)
          return binarySearch(arr, start, mid - 1, n, isStart, isEnd);
        else {
          return mid;
        }
      } else if (isEnd) {
        if (mid < arr.length - 1 && arr[mid + 1] == n)
          return binarySearch(arr, mid + 1, end, n, isStart, isEnd);
        else return mid;
      } else {
        return mid;
      }
    } else if (arr[mid] > n) {
      end = mid - 1;
    } else {
      start = mid + 1;
    }
  }
  return -1;
}

let arr = [1, 3, 5, 5, 5, 5, 7, 123, 125];
console.log(arr);
console.log(method1(arr, 0, arr.length - 1, 5));

File: SearchingAndSorting/LastAndFirst/LastAndFirst.go
package main

import "fmt"

func method1(arr []int, k int) []int {
	ind := binarySearch(arr, 0, len(arr)-1, k, false, false)
	left := binarySearch(arr, 0, ind, k, true, false)
	right := binarySearch(arr, ind, len(arr)-1, k, false, true)
	fmt.Println("ind is ", ind, ", left is ", left, ", right is ", right)
	return []int{left, right}
}

func binarySearch(arr []int, l int, r int, k int, isLeft bool, isRight bool) int {
	for l <= r {
		m := l + (r-l)/2
		if arr[m] == k {
			if isLeft {
				if m == 0 {
					return 0
				} else if arr[m] != arr[m-1] {
					return m
				} else {
					r = m - 1
				}
			} else if isRight {
				if m == r {
					return r - 1
				} else if arr[m] != arr[m+1] {
					return m
				} else {
					l = m + 1
				}
			} else {
				return m
			}
		} else if arr[m] > k {
			r = m - 1
		} else {
			l = m + 1
		}
	}
	return -1
}

func main() {
	arr := []int{1, 3, 5, 5, 5, 5, 67, 123}
	x := 5
	method1(arr, x)
}
